src/main/java/com/slack/kaldb/elasticsearchApi/ElasticsearchApiService.java

```diff
@@ -72,7 +73,30 @@ private EsSearchResponse doSearch(EsSearchRequest request) throws IOException {
     KaldbSearch.SearchRequest searchRequest = request.toKaldbSearchRequest();
     // TODO remove join when we move to query service
     List<SearchResponseHit> responseHits = new ArrayList<>();
-    KaldbSearch.SearchResult searchResult = searcher.doSearch(searchRequest).join();
+
+    CompletableFuture<KaldbSearch.SearchResult> searchResultFuture = new CompletableFuture<>();
+    StreamObserver<KaldbSearch.SearchResult> responseObserver =
+        new StreamObserver<>() {
+          private KaldbSearch.SearchResult searchResult;
+
+          @Override
+          public void onNext(KaldbSearch.SearchResult searchResult) {
+            this.searchResult = searchResult;
+          }
+
+          @Override
+          public void onError(Throwable throwable) {
+            searchResultFuture.completeExceptionally(throwable);
+          }
+
+          @Override
+          public void onCompleted() {
+            searchResultFuture.complete(searchResult);
+          }
+        };
+
+    searcher.search(searchRequest, responseObserver);
+    KaldbSearch.SearchResult searchResult = searchResultFuture.join();
```

Comment:
Can we avoid the `join` here and then make everything from here to transforming the results async?

Or is that not part of the scope here?

Reply:
For this PR I just did what was required to support both indexer & query use cases, as the query classes are currently under active development and are expected to change a bit more.

This use case is also a bit unique as this is all part of an http request/response cycle, which is why I also didn't put a timeout when resolving this future. I'm not sure having this written async is a huge benefit, given we need to return an http response at the end.